File: app/src/main/java/com/example/Maatran/MainActivity.java
package com.example.Maatran;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import com.example.Maatran.utils.UIFunctions;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.Objects;

public class MainActivity extends AppCompatActivity implements UIFunctions {

    private static final String TAG = "MainActivity";
    int lastUpdatedDot = 0;
    Handler loadingDotHandler;
    Runnable loadingDotRunnable = this::changeLoadingDot;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ConstraintLayout layout = findViewById(R.id.main_activity_bg);
        Drawable backgroundDrawable = layout.getBackground();
        changeStatusBarColor(backgroundDrawable,this);
        loadingDotHandler = new Handler();
        loadingDotHandler.postDelayed(loadingDotRunnable,300);
        signInOptions();
    }

    void changeLoadingDot()
    {
        String currLoadDotView = "load_dot"+ lastUpdatedDot % 3;
        String oldLoadDotView = "load_dot"+ (lastUpdatedDot % 3 == 0 ? 2 : lastUpdatedDot % 3 - 1);
        lastUpdatedDot++;
        int currResID = getResources().getIdentifier(currLoadDotView,"id",getPackageName());
        int oldResID = getResources().getIdentifier(oldLoadDotView,"id",getPackageName());
        AppCompatImageView currResView = findViewById(currResID);
        AppCompatImageView oldResView = findViewById(oldResID);
        Drawable drawable = AppCompatResources.getDrawable(this,R.drawable.loading_dot_grey);
        drawable.setColorFilter(ContextCompat.getColor(this,R.color.white), PorterDuff.Mode.SRC_IN);
        currResView.setImageDrawable(drawable);
        oldResView.setImageResource(R.drawable.loading_dot_grey);
        loadingDotHandler.postDelayed(loadingDotRunnable,300);
    }

    public void signInOptions() {
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        if(user==null) {
            Intent intent = new Intent(getApplicationContext(), WelcomeActivity.class);
            startActivity(intent);
            super.finish();
        }
        else
        {
            checkForUserDetails(user);
        }
    }

    private void checkForUserDetails(FirebaseUser user)
    {
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        db.collection("UserDetails").document(Objects.requireNonNull(user.getEmail())).get().addOnCompleteListener(task -> {
            if(task.isSuccessful())
            {
                DocumentSnapshot document = task.getResult();
                if(document.exists())
                {
                    if(document.get("name")==null)
                    {

                        Toast.makeText(this, "Fill in your details to proceed...", Toast.LENGTH_SHORT).show();
                        Intent intent = new Intent(getApplicationContext(), EditPatient.class);
                        intent.putExtra("isPatient",false);
                        intent.putExtra("newDetails", true);
                        startActivity(intent);
                        super.finish();
                    }
                    else
                    {
                        Intent intent = new Intent(getApplicationContext(), DashboardActivity.class);
                        startActivity(intent);
                        super.finish();
                    }
                }
                else {
                    Log.d(TAG, "Document does not exist");
                    Toast.makeText(this, "Please log in manually", Toast.LENGTH_SHORT).show();
                    Intent intent = new Intent(getApplicationContext(), WelcomeActivity.class);
                    startActivity(intent);
                    super.finish();
                }
            }
            else {
                Log.d(TAG, "Task failed to complete");
                Toast.makeText(this, "Please log in manually", Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(getApplicationContext(), WelcomeActivity.class);
                startActivity(intent);
                super.finish();
            }
        });
    }
}

File: app/src/main/java/com/example/Maatran/tests/MainActivity.kt
package com.example.Maatran.tests

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.Maatran.databinding.ActivityMainBinding
import com.example.Maatran.ui.EditPatientActivity
import com.example.Maatran.utils.commonUIFunctions
import com.example.Maatran.services.FirebaseServices

class MainActivity : AppCompatActivity(), commonUIFunctions {
    private lateinit var _binding: ActivityMainBinding

    private val _tag = "MainActivity"

    private var _lastUpdatedDot: Int = 0

    private lateinit var _loadingDotHandler: Handler

    private lateinit var _signInHandler: Handler

    private var _signInRunnable: Runnable = Runnable {
        Toast.makeText(this,"Its taking too long, please log-in/sign-up manually",Toast.LENGTH_SHORT).show()
        val intent = Intent(applicationContext, WelcomeActivity::class.java)
        startActivity(intent)
        super.finish()
    }
    private val _loadingDotRunnable: Runnable = Runnable { callLoadingDotAnim() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityMainBinding.inflate(layoutInflater)
        val rootView = _binding.root
        setContentView(rootView)
        _loadingDotHandler = Handler(Looper.getMainLooper())
        _loadingDotHandler.postDelayed(_loadingDotRunnable,500)
        checkForSignedInUser()
        _signInHandler = Handler(Looper.getMainLooper())
        _signInHandler.postDelayed(_signInRunnable,8000)
    }

    private fun callLoadingDotAnim()
    {
       _lastUpdatedDot =  changeLoadingDotAnimation(_lastUpdatedDot, this)
        _loadingDotHandler.postDelayed(_loadingDotRunnable, 500)
    }

    private fun checkForSignedInUser()
    {
        val firebaseServiceAuth: FirebaseServices.Auth = FirebaseServices.Auth{
            Log.v(_tag, it.toString())
            when(it)
            {
                1 -> {
                    val intent = Intent(applicationContext,AppNavigationActivity::class.java)
                    startActivity(intent)
                    super.finish()
                }
                0 -> {
                    Toast.makeText(
                        this,
                        "Fill in your details to proceed...",
                        Toast.LENGTH_SHORT
                    ).show()
                    val intent = Intent(applicationContext, EditPatientActivity::class.java)
                    intent.putExtra("isPatient", false)
                    intent.putExtra("newDetails", true)
                    startActivity(intent)
                    super.finish()
                }
                -1 -> {
                    val intent = Intent(applicationContext, WelcomeActivity::class.java)
                    startActivity(intent)
                    super.finish()
                }
                else -> {}
            }
        }
        firebaseServiceAuth.checkForExistingUser()
    }

    override fun onDestroy() {
        super.onDestroy()
        _signInHandler.removeCallbacks(_signInRunnable)
        _loadingDotHandler.removeCallbacks(_loadingDotRunnable)
    }
}